buildSrc/src/main/java/org/elasticsearch/gradle/JdkDownloadPlugin.java

```diff
@@ -105,87 +117,195 @@ private static void setupRootJdkDownload(Project rootProject, String platform, S
+                            TarArchiveEntry entry = tar.getNextTarEntry();
+                            while (entry != null) {
+                                /*
+                                 * We want to remove up to the and including the jdk-.* relative paths. That is a JDK archive is structured
+                                 * as:
+                                 *   jdk-12.0.1/
+                                 *   jdk-12.0.1/Contents
+                                 *   ...
+                                 *
+                                 * and we want to remove the leading jdk-12.0.1. Note however that there could also be a leading ./ as in
+                                 *   ./
+                                 *   ./jdk-12.0.1/
+                                 *   ./jdk-12.0.1/Contents
+                                 *
+                                 * so we account for this and search the path components until we find the jdk-12.0.1, and strip the leading
+                                 * components.
+                                 */
+                                final Path entryName = Paths.get(entry.getName());
+                                int index = 0;
+                                for (; index < entryName.getNameCount(); index++) {
+                                    if (entryName.getName(index).toString().matches("jdk-.*")) break;
+                                }
+                                if (index + 1 >= entryName.getNameCount()) {
+                                    // this happens on the top-level directories in the archive, which we are removing
+                                    entry = tar.getNextTarEntry();
+                                    continue;
+                                }
+                                // finally remove the top-level directories from the output path
+                                final Path destination = extractPath.resolve(entryName.subpath(index + 1, entryName.getNameCount()));
+                                final Path parent = destination.getParent();
+                                if (Files.exists(parent) == false) {
+                                    Files.createDirectories(parent);
+                                }
+                                if (entry.isDirectory()) {
+                                    Files.createDirectory(destination);
+                                } else if (entry.isSymbolicLink()) {
+                                    Files.createSymbolicLink(destination, Paths.get(entry.getLinkName()));
+                                } else {
+                                    // copy the file from the archive using a small buffer to avoid heaping
+                                    Files.createFile(destination);
+                                    tar.transferTo(new FileOutputStream(destination.toFile()));
```

Comment:
Constructing this `FileOutputStream` inline this way means it's never closed. We have to explicitly clean up here since `transferTo()` doesn't close this stream for us.

Reply:
I pushed 813a9bde0bffb280a600c2d5b722186369aa5e58.